src/storage-client/src/controller.rs

```diff
@@ -1738,170 +1757,143 @@ mod persist_read_handles {
-                                            {
-                                                let mut contents = Vec::new();
-                                                for ((source_data, _pid), _ts, diff) in read_handle
-                                                    .snapshot_and_fetch(as_of)
-                                                    .await
-                                                    .map_err(|_| StorageError::ReadBeforeSince(id))?
-                                                {
-                                                    let row = source_data.expect("cannot read snapshot").0?;
-                                                    contents.push((row, diff));
-                                                }
+                                        continue;
+                                    }
+                                };
 
-                                                Ok(contents)
-                                            }
+                                let mut persist_clients = persist_clients.lock().await;
 
-                                            let result = match read_handles.get_mut(&id) {
-                                                Some(Some(read_handle)) => {
-                                                    snapshot(read_handle, id, as_of)
-                                                    .instrument(span.clone())
-                                                    .await
-                                                },
-                                                Some(None) | None => {
-                                                    // A Some(None) means we downgraded since
-                                                    // to empty antichain (aka we dropped it).
-                                                    Err(StorageError::IdentifierMissing(id))
-                                                }
-                                            };
-                                            oneshot.send(result).expect("Oneshot should not fail");
-                                        }
-                                        PersistWorkerCmd::Shutdown => {
-                                            shutdown = true;
-                                        }
-                                    }
-                                }
+                                let persist_client = persist_clients
+                                    .open(metadata.persist_location.clone())
+                                    .await
+                                    .unwrap();
 
-                                let mut drops = HashSet::new();
-                                if !downgrades.is_empty() {
-                                    let futs = FuturesUnordered::new();
+                                // We'll register a new read handle, do the snapshot, and
```

Comment:
nit: I'd mention _why_ (so we don't have to keep heartbeating a handle just to occasionally use it for a snapshot)